### PR TITLE
Feature/add resource tags for appmesh mesh resource

### LINKED ISSUE #8101: AWS App Mesh tagging
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

As part of AWS App Mesh [GA](https://aws.amazon.com/about-aws/whats-new/2019/03/aws-app-mesh-is-now-generally-available/), resource tagging is [enabled](https://github.com/aws/aws-app-mesh-roadmap/issues/19).

### PR DESCRIPTION
**Pull request by kterada0509**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Reference #8101 

Changes proposed in this pull request:

* Add tag resource for appmesh service
* Add tags attribute for `aws_appmesh_mesh` resource

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSAppmesh/Mesh'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSAppmesh/Mesh -timeout 120m
=== RUN   TestAccAWSAppmesh
=== RUN   TestAccAWSAppmesh/Mesh
=== RUN   TestAccAWSAppmesh/Mesh/tags
=== RUN   TestAccAWSAppmesh/Mesh/basic
--- PASS: TestAccAWSAppmesh (108.16s)
    --- PASS: TestAccAWSAppmesh/Mesh (108.16s)
        --- PASS: TestAccAWSAppmesh/Mesh/tags (77.71s)
        --- PASS: TestAccAWSAppmesh/Mesh/basic (30.45s)
PASS
ok  	github.com/terraform-providers/tmp-terraform-provider-aws/aws	108.233s
```

### CODE REVIEW

**Comment by kterada0509** [comment]
Output from acceptance testing:
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSAppmesh/Mesh'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSAppmesh/Mesh -timeout 120m
=== RUN   TestAccAWSAppmesh
=== RUN   TestAccAWSAppmesh/Mesh
=== RUN   TestAccAWSAppmesh/Mesh/basic
=== RUN   TestAccAWSAppmesh/Mesh/egressFilter
=== RUN   TestAccAWSAppmesh/Mesh/tags
--- PASS: TestAccAWSAppmesh (179.38s)
    --- PASS: TestAccAWSAppmesh/Mesh (179.38s)
        --- PASS: TestAccAWSAppmesh/Mesh/basic (37.31s)
        --- PASS: TestAccAWSAppmesh/Mesh/egressFilter (67.56s)
        --- PASS: TestAccAWSAppmesh/Mesh/tags (74.51s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	179.459s
```

**Comment by ewbankkit** [comment]
Can we get a review of this please?
`tagsAppmesh.go` is required for adding resource tags to the other App Mesh resources listed in #8101.

File: aws/resource_aws_appmesh_mesh.go [hunk 1]
```diff
@@ -127,6 +130,15 @@ func resourceAwsAppmeshMeshRead(d *schema.ResourceData, meta interface{}) error
 		return fmt.Errorf("error setting spec: %s", err)
 	}
 
+	if err := saveTagsAppmesh(conn, d, aws.StringValue(resp.Mesh.Metadata.Arn)); err != nil {
```
Comment (ewbankkit) [inline aws/resource_aws_appmesh_mesh.go]: I think the preferred style for error handling is

```go
err = saveTagsAppmesh(conn, d, aws.StringValue(resp.Mesh.Metadata.Arn))
if isAWSErr(err, appmesh.ErrCodeNotFoundException, "") {
    log.Printf("[WARN] App Mesh service mesh (%s) not found, removing from state", d.Id())
    d.SetId("")
    return nil
}
if err != nil {
    return fmt.Errorf("error saving tags: %s", err)
}
```

File: aws/resource_aws_appmesh_mesh.go [hunk 2]
```diff
@@ -147,6 +159,15 @@ func resourceAwsAppmeshMeshUpdate(d *schema.ResourceData, meta interface{}) erro
 		}
 	}
 
+	if err := setTagsAppmesh(conn, d, d.Get("arn").(string)); err != nil {
```
Comment (ewbankkit) [inline aws/resource_aws_appmesh_mesh.go]: Same comment as [above](https://github.com/terraform-providers/terraform-provider-aws/pull/8111/files#r277181257).

**Comment by kterada0509** [comment]
Re-run tests.
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSAppmesh/Mesh'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSAppmesh/Mesh -timeout 120m
=== RUN   TestAccAWSAppmesh
=== RUN   TestAccAWSAppmesh/Mesh
=== RUN   TestAccAWSAppmesh/Mesh/basic
=== RUN   TestAccAWSAppmesh/Mesh/egressFilter
=== RUN   TestAccAWSAppmesh/Mesh/tags
--- PASS: TestAccAWSAppmesh (185.23s)
    --- PASS: TestAccAWSAppmesh/Mesh (185.23s)
        --- PASS: TestAccAWSAppmesh/Mesh/basic (36.70s)
        --- PASS: TestAccAWSAppmesh/Mesh/egressFilter (68.07s)
        --- PASS: TestAccAWSAppmesh/Mesh/tags (80.46s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	185.314s
```

File: aws/resource_aws_appmesh_mesh.go (hunk 1 above)
Comment (kterada0509) [reply aws/resource_aws_appmesh_mesh.go]: Fixed.

File: aws/resource_aws_appmesh_mesh.go (hunk 2 above)
Comment (kterada0509) [reply aws/resource_aws_appmesh_mesh.go]: Fixed too.

**Comment by kterada0509** [comment]
@ewbankkit could you review fixes?

**Comment by ewbankkit** [comment]
```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSAppmesh/Mesh'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSAppmesh/Mesh -timeout 120m
=== RUN   TestAccAWSAppmesh
=== RUN   TestAccAWSAppmesh/Mesh
=== RUN   TestAccAWSAppmesh/Mesh/egressFilter
=== RUN   TestAccAWSAppmesh/Mesh/tags
=== RUN   TestAccAWSAppmesh/Mesh/basic
--- PASS: TestAccAWSAppmesh (115.91s)
    --- PASS: TestAccAWSAppmesh/Mesh (115.91s)
        --- PASS: TestAccAWSAppmesh/Mesh/egressFilter (44.72s)
        --- PASS: TestAccAWSAppmesh/Mesh/tags (50.00s)
        --- PASS: TestAccAWSAppmesh/Mesh/basic (21.19s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	115.996s
```

File: aws/resource_aws_appmesh_mesh.go [hunk 3]
```diff
@@ -157,7 +180,7 @@ func resourceAwsAppmeshMeshDelete(d *schema.ResourceData, meta interface{}) erro
 	_, err := conn.DeleteMesh(&appmesh.DeleteMeshInput{
 		MeshName: aws.String(d.Id()),
 	})
-	if isAWSErr(err, "NotFoundException", "") {
+	if isAWSErr(err, appmesh.ErrCodeNotFoundException, "") {
```
Comment (ewbankkit) [inline aws/resource_aws_appmesh_mesh.go]: Nice. Could you also replace the hard-coded string value in `resourceAwsAppmeshMeshRead()`? https://github.com/terraform-providers/terraform-provider-aws/blob/1bb0dead12829bf6c9da9e3170f52a8f24984497/aws/resource_aws_appmesh_mesh.go#L107

File: aws/tagsAppmesh.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+// be destroyed.
+func diffTagsAppmesh(oldTags, newTags []*appmesh.TagRef) ([]*appmesh.TagRef, []*string) {
+	// First, we're creating everything we have
+	create := make(map[string]interface{})
+	for _, t := range newTags {
+		create[*t.Key] = *t.Value
+	}
+
+	// Build the list of what to remove
+	var remove []*string
```
Comment (ewbankkit) [inline aws/tagsAppmesh.go]: Need to handle the overlapping tags case, e.g. from ECR https://github.com/terraform-providers/terraform-provider-aws/blob/1bb0dead12829bf6c9da9e3170f52a8f24984497/aws/tagsECR.go#L79-L89

See https://github.com/terraform-providers/terraform-provider-aws/issues/7052.

File: aws/tagsAppmesh.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+		dt = resp.Tags
+	}
+
+	return d.Set("tags", tagsToMapAppmesh(dt))
+}
+
+// compare a tag against a list of strings and checks if it should
+// be ignored or not
+func tagIgnoredAppmesh(t *appmesh.TagRef) bool {
+	filter := []string{"^aws:", "^appmesh:", "Name"}
```
Comment (ewbankkit) [inline aws/tagsAppmesh.go]: Is there a need to exclude `Name` and `appmesh:` here?
In my testing these tags don't seem to be added automagically and I can add manually:
```console
$ aws --region us-west-2 appmesh tag-resource --resource-arn "arn:aws:appmesh:us-west-2:000000000000:mesh/simpleapp" --tags "key=Name,value=test"
$ aws --region us-west-2 appmesh tag-resource --resource-arn "arn:aws:appmesh:us-west-2:000000000000:mesh/simpleapp" --tags "key=appmesh:k,value=v"
$ aws --region us-west-2 appmesh list-tags-for-resource --resource-arn "arn:aws:appmesh:us-west-2:000000000000:mesh/simpleapp"
{
    "tags": [
        {
            "value": "test", 
            "key": "Name"
        }, 
        {
            "value": "v", 
            "key": "appmesh:k"
        }
    ]
}
```

**Comment by kterada0509** [comment]
Re-run acctest.

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSAppmesh/Mesh'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSAppmesh/Mesh -timeout 120m
=== RUN   TestAccAWSAppmesh
=== RUN   TestAccAWSAppmesh/Mesh
=== RUN   TestAccAWSAppmesh/Mesh/basic
=== RUN   TestAccAWSAppmesh/Mesh/egressFilter
=== RUN   TestAccAWSAppmesh/Mesh/tags
--- PASS: TestAccAWSAppmesh (142.94s)
    --- PASS: TestAccAWSAppmesh/Mesh (142.94s)
        --- PASS: TestAccAWSAppmesh/Mesh/basic (27.22s)
        --- PASS: TestAccAWSAppmesh/Mesh/egressFilter (52.34s)
        --- PASS: TestAccAWSAppmesh/Mesh/tags (63.38s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	143.056s
```

File: aws/tagsAppmesh.go (hunk 5 above)
Comment (kterada0509) [reply aws/tagsAppmesh.go]: Thanks, I remove `Name` and `appmesh:` from ignore list

File: aws/tagsAppmesh.go (hunk 4 above)
Comment (kterada0509) [reply aws/tagsAppmesh.go]: Fixed it.

File: aws/resource_aws_appmesh_mesh.go (hunk 3 above)
Comment (kterada0509) [reply aws/resource_aws_appmesh_mesh.go]: Fixed it.

**Comment by bflad** [comment]
(Fixing the minor unit/linting error on merge so we can get this released today)

```
aws/tagsAppmesh_test.go:29:4: cannot use map[string]string literal (type map[string]string) as type []string in field value
aws/tagsAppmesh_test.go:43:4: cannot use map[string]string literal (type map[string]string) as type []string in field value
aws/tagsAppmesh_test.go:61:4: cannot use map[string]string literal (type map[string]string) as type []string in field value
aws/tagsAppmesh_test.go:76:4: cannot use map[string]string literal (type map[string]string) as type []string in field value
```

**Comment by bflad** [comment]
This has been released in [version 2.17.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2170-june-28-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

For further feature requests or bug reports with this functionality, please create a [new GitHub issue](https://github.com/terraform-providers/terraform-provider-aws/issues/new/choose) following the template for triage. Thanks!

**Pull request merged**